File: languages/clojure/exercises/concept/numbers/test/numbers_test.clj
(ns numbers-test
  (:require [clojure.test :refer [deftest testing is]]
            numbers))

(deftest production-rate-test
  (testing "Production rate for"
    (testing "speed 0"
      (is (= 0.0 (numbers/production-rate 0))))
    (testing "speed 1"
      (is (= 221.0 (numbers/production-rate 1))))
    (testing "speed 4"
      (is (= 884.0 (numbers/production-rate 4))))
    (testing "speed 7"
      (is (= 1392.3 (numbers/production-rate 7))))
    (testing "speed 9"
      (is (= 1591.2 (numbers/production-rate 9))))
    (testing "speed 10"
      (is (= 1701.7 (numbers/production-rate 10))))))

(deftest working-items-test
  (testing "Working items for"
    (testing "speed 0"
      (is (= 0 (numbers/working-items 0))))
    (testing "speed 1"
      (is (= 3 (numbers/working-items 1))))
    (testing "speed 5"
      (is (= 16 (numbers/working-items 5))))
    (testing "speed 8"
      (is (= 26 (numbers/working-items 8))))
    (testing "speed 9"
      (is (= 26 (numbers/working-items 9))))
    (testing "speed 10"
      (is (= 28 (numbers/working-items 10))))))

File: languages/cpp/exercises/concept/numbers/numbers_test.cpp
#include "numbers.h"

#include "test/catch.hpp"

constexpr auto margin = 1e-5;

TEST_CASE("Production rate per hour for speed zero") {
    auto rate = assembly_line::production_rate_per_hour(0);
    static_assert(std::is_same<double, decltype(rate)>::value,
                  "Returned value should be of type double");
    REQUIRE(Approx(0.0).margin(margin) == rate);
}

#if defined(EXERCISM_RUN_ALL_TESTS)
TEST_CASE("Production rate_per_hour_for_speed_one") {
    auto rate = assembly_line::production_rate_per_hour(1);
    static_assert(std::is_same<double, decltype(rate)>::value,
                  "Returned value should be of type double");
    REQUIRE(Approx(221.0).margin(margin) == rate);
}

TEST_CASE("Production rate_per_hour_for_speed_four") {
    auto rate = assembly_line::production_rate_per_hour(4);
    static_assert(std::is_same<double, decltype(rate)>::value,
                  "Returned value should be of type double")
        REQUIRE(Approx(884.0).margin(margin) == rate);
}

TEST_CASE("Production rate_per_hour_for_speed_seven") {
    auto rate = assembly_line::production_rate_per_hour(7);
    static_assert(std::is_same<double, decltype(rate)>::value,
                  "Returned value should be of type double");
    REQUIRE(Approx(1392.3).margin(margin) == rate);
}

TEST_CASE("Working items per minute for speed zero") {
    auto rate = assembly_line::working_items_per_minute(0);
    static_assert(std::is_same<std::int32_t, decltype(rate)>::value,
                  "Returned value should be of type std::int32_t");
    REQUIRE(0 == rate);
}

TEST_CASE("Working items per minute for speed one") {
    auto rate = assembly_line::working_items_per_minute(1);
    static_assert(std::is_same<std::int32_t, decltype(rate)>::value,
                  "Returned value should be of type std::int32_t");
    REQUIRE(3 == rate);
}

TEST_CASE("Working items per minute for speed eight") {
    auto rate = assembly_line::working_items_per_minute(8);
    static_assert(std::is_same<std::int32_t, decltype(rate)>::value,
                  "Returned value should be of type std::int32_t");
    REQUIRE(26 == rate);
}

TEST_CASE("Working items per minute for speed nine") {
    auto rate = assembly_line::working_items_per_minute(9);
    static_assert(std::is_same<std::int32_t, decltype(rate)>::value,
                  "Returned value should be of type std::int32_t");
    REQUIRE(26 == rate);
}

TEST_CASE("Working items per minute for speed ten") {
    auto rate = assembly_line::working_items_per_minute(10);
    static_assert(std::is_same<std::int32_t, decltype(rate)>::value,
                  "Returned value should be of type std::int32_t");
    REQUIRE(28 == rate);
}
#endif
